conformance-tests/src/main.rs

```diff
@@ -0,0 +1,257 @@
+        .unwrap();
+        for invocation in test.config.invocations {
+            let conformance_tests::config::Invocation::Http(mut invocation) = invocation;
+            invocation.request.substitute_from_env(&mut env).unwrap();
+            let shim = env.runtime_mut();
+            if let Err(e) = invocation.run(|request| shim.make_http_request(request)) {
+                println!("❌ test failed: {}", test.name);
+                println!("error: {}", e);
+                for e in e.chain() {
+                    println!("\t{}", e);
+                }
+
+                continue 'test;
+            }
+        }
+        println!("✅ test passed: {}", test.name);
+    }
+}
+
+struct SpinShim {
+    process: std::process::Child,
+    #[allow(dead_code)]
+    stdout: OutputStream,
+    stderr: OutputStream,
+    io_mode: IoMode,
+}
+
+/// `ctr run` invocations require an ID that is unique to all currently running instances. Since
+/// only one test runs at a time, we can reuse a constant ID.
+const CTR_RUN_ID: &str = "run-id";
+
+impl SpinShim {
+    pub fn config(
+        ctr_binary: PathBuf,
+        spin_binary: PathBuf,
+        oci_image: String,
+        services_config: ServicesConfig,
+    ) -> TestEnvironmentConfig<SpinShim> {
+        TestEnvironmentConfig {
+            services_config,
+            create_runtime: Box::new(move |env| {
+                let oci_image = format!(
+                    "localhost:{port}/{oci_image}:latest",
+                    port = env
+                        .get_port(5000)?
+                        .context("environment doesn't expose port for OCI registry")?
+                );
+                SpinShim::regisry_push(&spin_binary, &oci_image, env)?;
+                SpinShim::image_pull(&ctr_binary, &oci_image)?;
+                SpinShim::start(&ctr_binary, env, &oci_image, CTR_RUN_ID)
+            }),
+        }
+    }
+
+    pub fn regisry_push<R>(
+        spin_binary_path: &Path,
+        image: &str,
+        env: &mut TestEnvironment<R>,
+    ) -> anyhow::Result<()> {
+        // TODO: consider enabling configuring a port
+        let mut cmd = Command::new(spin_binary_path);
+        cmd.args(["registry", "push", "-k"]).arg(image);
+        env.run_in(&mut cmd)
+            .context("failed to push spin app to registry with 'spin'")?;
+        Ok(())
+    }
+
+    pub fn image_pull(ctr_binary_path: &Path, image: &str) -> anyhow::Result<()> {
+        // TODO: consider enabling configuring a port
+        Command::new(ctr_binary_path)
+            .args(["image", "pull"])
+            .arg(image)
+            .output()
+            .context("failed to pull spin app with 'ctr'")?;
+        // TODO: assess output
+        Ok(())
+    }
+
+    /// Start the Spin app using `ctr run`
+    pub fn start<R>(
+        ctr_binary_path: &Path,
+        env: &mut TestEnvironment<R>,
+        image: &str,
+        ctr_run_id: &str,
+    ) -> anyhow::Result<Self> {
+        let port = 80;
```

Comment:
Currently this is the default (hard-coded) port for HTTP applications; is the intention that this would be configurable for the tests (ref #138)?

Reply:
This could be configured now; however, just using a fixed port for the tests is fine for now given that one test is run at a time. Left a todo to consider making it configurable

Reply:
Now the port configuration is a possible, I would suggest we try to run on a free port. My Linux box runs a webserver on port 80 and I have to remember to turn it off before being able to run these tests. I imagine I'm not alone in having something already listening on port 80.

The Spin CLI conformance tests [look for a random free port](https://github.com/fermyon/spin/blob/444adbaac3271a358da560c077d2c25c756b3056/tests/testing-framework/src/runtimes/spin_cli.rs#L258-L262) to bind to, and it would be nice if we could do the same here.

Reply:
Added! We are now randomly selecting a port and configuring with `SPIN_HTTP_LISTEN_ADDR`